google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher.java

```diff
@@ -426,21 +431,37 @@ int computeDeadlineSeconds() {
 
   @InternalApi
   void extendDeadlines() {
-    List<String> acksToSend = Collections.<String>emptyList();
-    PendingModifyAckDeadline modack = new PendingModifyAckDeadline(getMessageDeadlineSeconds());
+    int extendSeconds = getMessageDeadlineSeconds();
+    PendingModifyAckDeadline modack = new PendingModifyAckDeadline(extendSeconds);
+    ArrayListMultimap<Long, String> expiringModack = ArrayListMultimap.create();
     Instant now = now();
+    Instant extendTo = now.plusSeconds(extendSeconds);
 
     Iterator<Map.Entry<String, Instant>> it = pendingMessages.entrySet().iterator();
     while (it.hasNext()) {
       Map.Entry<String, Instant> entry = it.next();
-      if (entry.getValue().isBefore(now)) {
-        it.remove();
-      } else {
-        modack.ackIds.add(entry.getKey());
+      String ackId = entry.getKey();
+      Instant totalExpiration = entry.getValue();
+      if (totalExpiration.isAfter(extendTo)) {
+        // Most extensions won't reach total expiration, so avoid map lookup.
+        modack.ackIds.add(ackId);
+        continue;
       }
+      it.remove();
+      if (totalExpiration.isAfter(now)) {
+        expiringModack.put(now.until(totalExpiration, ChronoUnit.SECONDS), ackId);
```

Comment:
Maybe we should have some threshold here, if it's 10 milliseconds, we are going to nack the message, which doesn't seem quite right. We should probably mod ack either the max of 1 second and this value, or just ignore it if it is going to expire within a second anyway.

Reply:
Done